### PR TITLE
Allow overriding project directory with WHIPPET_PROJECT_DIRECTORY

### PR DESCRIPTION
**Pull request by tomdxw**

### CODE REVIEW

File: tests/project_directory_test.php [hunk 1]
```diff
@@ -107,4 +107,20 @@ public function testGetDirectoryAvoidPluginsDirectory()
         $this->assertInstanceOf('\\Dxw\\Whippet\\ProjectDirectory', $result->unwrap());
         $this->assertEquals($dir, $result->unwrap()->__toString());
     }
+
+    public function testGetDirectoryWithEnv()
+    {
+        $dir = $this->getDir();
+
+        putenv('WHIPPET_PROJECT_DIRECTORY=/etc');
```
Comment (dgmstuart) [inline tests/project_directory_test.php]: Do we need to clean up afterwards? Will this value be set for all subsequent tests?

Is there any risk that we could end up writing test data into `/etc` which we leave lying around after the test suite run?

File: src/ProjectDirectory.php [hunk 2]
```diff
@@ -6,6 +6,11 @@ class ProjectDirectory
 {
     public static function find(/* string */ $cwd)
     {
+        $env = getenv('WHIPPET_PROJECT_DIRECTORY');
```
Comment (dgmstuart) [inline src/ProjectDirectory.php]: Would it be useful to document this somewhere?

File: src/ProjectDirectory.php [hunk 3]
```diff
@@ -6,6 +6,11 @@ class ProjectDirectory
...
+        if ($env) {
+            return \Result\Result::ok(new self($env));
```
Comment (dgmstuart) [inline src/ProjectDirectory.php]: In the normal case we check for whippet.json - do we not care if it doesn't exist?

File: tests/project_directory_test.php (hunk 1 above)
Comment (tomdxw) [inline tests/project_directory_test.php]: Added a tearDown() to remove the env var.

**Comment by tomdxw** [comment]
> Looks sensible, but it would be really useful to have a note in the commit message explaining where this is useful - presumably you have a use case that has prompted this?

Mentioned the rationale in the commit message. Specifically, I was going to use it in https://github.com/dxw/wpc

File: src/ProjectDirectory.php (hunk 3 above)
Comment (tomdxw) [inline src/ProjectDirectory.php]: Good question. I was thinking of this as an override. So you can just force whippet to use a particular directory without it asking any questions.

But I suppose I should probably make it use the normal logic.

Comment (dgmstuart) [inline src/ProjectDirectory.php]: Up to you - I genuinely don't know which is preferable - depends on use case I guess?

File: tests/project_directory_test.php (hunk 1 above)
Comment (dgmstuart) [inline tests/project_directory_test.php]: Another thought: would this override the user's local environment? So if someone had this var set up for their local dev (probably unlikely) then the test would override that? 

Possibly something to file under 'maybe possible, but not likely enough to worry about'

Comment (dgmstuart) [inline tests/project_directory_test.php]: Out of curiosity: this looks like it should leave files hanging around in my /etc directory, but it doesn't - how does that work?

Normally for this sort of thing I'd isolate it by creating a `tests/temp` directory and deleting that at the end, but whatever works.

Comment (tomdxw) [inline tests/project_directory_test.php]: The ProjectDirectory class doesn't modify any files. And the `mkdir` and `touch` calls below are operating on files within `vfs://root`. PHP has (partial) support for virtual filesystems. In this case we're using [vfsStream](https://github.com/mikey179/vfsStream) which is an in-memory filesystem meant to be used with tests.

Comment (dgmstuart) [inline tests/project_directory_test.php]: Ah cool. I'd seen the use of vfs in php tests before, but only explicitly. It's cool that you can do this transparently.

**Comment by dgmstuart** [comment]
@tomdxw this is nearly good to go - the only thing it needs is documenting somewhere - probably in the README?

**Comment by tomdxw** [comment]
I haven't used whippet since `wpc` was invented. Since nobody was using it previously, we can probably safely call `whippet server` deprecated and close any pull requests/issues for it.

**Pull request closed without merging**